// @stripe/agent-toolkit/cloudflare/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/cloudflare/index.ts
var index_exports = {};
__export(index_exports, {
  experimental_PaidMcpAgent: () => experimental_PaidMcpAgent
});
module.exports = __toCommonJS(index_exports);

// src/modelcontextprotocol/register-paid-tool.ts
var import_stripe = __toESM(require("stripe"));
async function registerPaidTool(mcpServer, toolName, toolDescription, paramsSchema, paidCallback, options) {
  const priceId = options.checkout.line_items?.find((li) => li.price)?.price;
  if (!priceId) {
    throw new Error(
      "Price ID is required for a paid MCP tool. Learn more about prices: https://docs.stripe.com/products-prices/how-products-and-prices-work"
    );
  }
  const stripe = new import_stripe.default(options.stripeSecretKey, {
    appInfo: {
      name: "stripe-agent-toolkit-mcp-payments",
      version: "0.7.11",
      url: "https://github.com/stripe/agent-toolkit"
    }
  });
  const getCurrentCustomerID = async () => {
    const customers = await stripe.customers.list({
      email: options.userEmail
    });
    let customerId = null;
    if (customers.data.length > 0) {
      customerId = customers.data.find((customer) => {
        return customer.email === options.userEmail;
      })?.id || null;
    }
    if (!customerId) {
      const customer = await stripe.customers.create({
        email: options.userEmail
      });
      customerId = customer.id;
    }
    return customerId;
  };
  const isToolPaidFor = async (toolName2, customerId) => {
    const sessions = await stripe.checkout.sessions.list({
      customer: customerId,
      limit: 100
    });
    const paidSession = sessions.data.find(
      (session) => session.metadata?.toolName === toolName2 && session.payment_status === "paid"
    );
    if (paidSession?.subscription) {
      const subs = await stripe.subscriptions.list({
        customer: customerId || "",
        status: "active"
      });
      const activeSub = subs.data.find(
        (sub) => sub.items.data.find((item) => item.price.id === priceId)
      );
      if (activeSub) {
        return true;
      }
    }
    if (paidSession) {
      return true;
    }
    return false;
  };
  const createCheckoutSession = async (paymentType, customerId) => {
    try {
      const session = await stripe.checkout.sessions.create({
        ...options.checkout,
        metadata: {
          ...options.checkout.metadata,
          toolName
        },
        customer: customerId || void 0
      });
      const result = {
        status: "payment_required",
        data: {
          paymentType,
          checkoutUrl: session.url,
          paymentReason: options.paymentReason
        }
      };
      return {
        content: [
          {
            type: "text",
            text: JSON.stringify(result)
          }
        ]
      };
    } catch (error) {
      let errMsg = "Unknown error";
      if (typeof error === "object" && error !== null) {
        if ("raw" in error && typeof error.raw?.message === "string") {
          errMsg = error.raw.message;
        } else if ("message" in error && typeof error.message === "string") {
          errMsg = error.message;
        }
      }
      console.error("Error creating stripe checkout session", errMsg);
      return {
        content: [
          {
            type: "text",
            text: JSON.stringify({
              status: "error",
              error: errMsg
            })
          }
        ],
        isError: true
      };
    }
  };
  const recordUsage = async (customerId) => {
    if (!options.meterEvent) return;
    await stripe.billing.meterEvents.create({
      event_name: options.meterEvent,
      payload: {
        stripe_customer_id: customerId,
        value: "1"
      }
    });
  };
  const callback = async (args, extra) => {
    const customerId = await getCurrentCustomerID();
    const paidForTool = await isToolPaidFor(toolName, customerId);
    const paymentType = options.meterEvent ? "usageBased" : "oneTimeSubscription";
    if (!paidForTool) {
      const checkoutResult = await createCheckoutSession(
        paymentType,
        customerId
      );
      if (checkoutResult) return checkoutResult;
    }
    if (paymentType === "usageBased") {
      await recordUsage(customerId);
    }
    return paidCallback(args, extra);
  };
  mcpServer.tool(toolName, toolDescription, paramsSchema, callback);
  await Promise.resolve();
}

// src/cloudflare/index.ts
var import_mcp = require("agents/mcp");
var experimental_PaidMcpAgent = class extends import_mcp.McpAgent {
  paidTool(toolName, toolDescription, paramsSchema, paidCallback, options) {
    const mcpServer = this.server;
    const userEmail = this.props.userEmail;
    const updatedOptions = {
      ...options,
      userEmail,
      // @ts-ignore
      stripeSecretKey: this.env.STRIPE_SECRET_KEY
    };
    registerPaidTool(
      mcpServer,
      toolName,
      toolDescription,
      paramsSchema,
      paidCallback,
      updatedOptions
    );
  }
};
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  experimental_PaidMcpAgent
});
