// @stripe/agent-toolkit/langchain/index.js
"use strict";
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  // If the importer is in node compatibility mode or this is not an ESM
  // file that has been converted to a CommonJS file using a Babel-
  // compatible transform (i.e. "__esModule" has not been set), then set
  // "default" to the CommonJS "module.exports" for node compatibility.
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/langchain/index.ts
var index_exports = {};
__export(index_exports, {
  StripeAgentToolkit: () => toolkit_default
});
module.exports = __toCommonJS(index_exports);

// src/langchain/tool.ts
var import_tools = require("@langchain/core/tools");
var StripeTool = class extends import_tools.StructuredTool {
  stripeAPI;
  method;
  name;
  description;
  schema;
  constructor(StripeAPI2, method, description, schema) {
    super();
    this.stripeAPI = StripeAPI2;
    this.method = method;
    this.name = method;
    this.description = description;
    this.schema = schema;
  }
  _call(arg, _runManager, _parentConfig) {
    return this.stripeAPI.run(this.method, arg);
  }
};
var tool_default = StripeTool;

// src/shared/api.ts
var import_stripe = __toESM(require("stripe"));

// src/shared/customers/createCustomer.ts
var import_zod = require("zod");
var createCustomerPrompt = (_context = {}) => `
This tool will create a customer in Stripe.

It takes two arguments:
- name (str): The name of the customer.
- email (str, optional): The email of the customer.
`;
var createCustomerParameters = (_context = {}) => import_zod.z.object({
  name: import_zod.z.string().describe("The name of the customer"),
  email: import_zod.z.string().email().optional().describe("The email of the customer")
});
var createCustomerAnnotations = () => ({
  destructiveHint: false,
  idempotentHint: false,
  openWorldHint: true,
  readOnlyHint: false,
  title: "Create customer"
});
var createCustomer = async (stripe, context, params) => {
  try {
    const customer = await stripe.customers.create(
      params,
      context.account ? { stripeAccount: context.account } : void 0
    );
    return { id: customer.id };
  } catch (error) {
    return "Failed to create customer";
  }
};
var tool = (context) => ({
  method: "create_customer",
  name: "Create Customer",
  description: createCustomerPrompt(context),
  parameters: createCustomerParameters(context),
  annotations: createCustomerAnnotations(),
  actions: {
    customers: {
      create: true
    }
  },
  execute: createCustomer
});
var createCustomer_default = tool;

// src/shared/customers/listCustomers.ts
var import_zod2 = require("zod");
var listCustomersPrompt = (_context = {}) => `
This tool will fetch a list of Customers from Stripe.

It takes two arguments:
- limit (int, optional): The number of customers to return.
- email (str, optional): A case-sensitive filter on the list based on the customer's email field.
`;
var listCustomersParameters = (_context = {}) => import_zod2.z.object({
  limit: import_zod2.z.number().int().min(1).max(100).optional().describe(
    "A limit on the number of objects to be returned. Limit can range between 1 and 100."
  ),
  email: import_zod2.z.string().optional().describe(
    "A case-sensitive filter on the list based on the customer's email field. The value must be a string."
  )
});
var listCustomersAnnotations = () => ({
  destructiveHint: false,
  idempotentHint: true,
  openWorldHint: true,
  readOnlyHint: true,
  title: "List customers"
});
var listCustomers = async (stripe, context, params) => {
  try {
    const customers = await stripe.customers.list(
      params,
      context.account ? { stripeAccount: context.account } : void 0
    );
    return customers.data.map((customer) => ({ id: customer.id }));
  } catch (error) {
    return "Failed to list customers";
  }
};
var tool2 = (context) => ({
  method: "list_customers",
  name: "List Customers",
  description: listCustomersPrompt(context),
  parameters: listCustomersParameters(context),
  annotations: listCustomersAnnotations(),
  actions: {
    customers: {
      read: true
    }
  },
  execute: listCustomers
});
var listCustomers_default = tool2;

// src/shared/products/createProduct.ts
var import_zod3 = require("zod");
var createProductPrompt = (_context = {}) => `
This tool will create a product in Stripe.

It takes two arguments:
- name (str): The name of the product.
- description (str, optional): The description of the product.
`;
var createProduct = async (stripe, context, params) => {
  try {
    const product = await stripe.products.create(
      params,
      context.account ? { stripeAccount: context.account } : void 0
    );
    return product;
  } catch (error) {
    return "Failed to create product";
  }
};
var createProductParameters = (_context = {}) => import_zod3.z.object({
  name: import_zod3.z.string().describe("The name of the product."),
  description: import_zod3.z.string().optional().describe("The description of the product.")
});
var createProductAnnotations = () => ({
  destructiveHint: false,
  idempotentHint: false,
  openWorldHint: true,
  readOnlyHint: false,
  title: "Create product"
});
var tool3 = (context) => ({
  method: "create_product",
  name: "Create Product",
  description: createProductPrompt(context),
  parameters: createProductParameters(context),
  annotations: createProductAnnotations(),
  actions: {
    products: {
      create: true
    }
  },
  execute: createProduct
});
var createProduct_default = tool3;

// src/shared/products/listProducts.ts
var import_zod4 = require("zod");
var listProductsPrompt = (_context = {}) => `
This tool will fetch a list of Products from Stripe.

It takes one optional argument:
- limit (int, optional): The number of products to return.
`;
var listProducts = async (stripe, context, params) => {
  try {
    const products = await stripe.products.list(
      params,
      context.account ? { stripeAccount: context.account } : void 0
    );
    return products.data;
  } catch (error) {
    return "Failed to list products";
  }
};
var listProductsParameters = (_context = {}) => import_zod4.z.object({
  limit: import_zod4.z.number().int().min(1).max(100).optional().describe(
    "A limit on the number of objects to be returned. Limit can range between 1 and 100, and the default is 10."
  )
});
var listProductsAnnotations = () => ({
  destructiveHint: false,
  idempotentHint: true,
  openWorldHint: true,
  readOnlyHint: true,
  title: "List products"
});
var tool4 = (context) => ({
  method: "list_products",
  name: "List Products",
  description: listProductsPrompt(context),
  parameters: listProductsParameters(context),
  annotations: listProductsAnnotations(),
  actions: {
    products: {
      read: true
    }
  },
  execute: listProducts
});
var listProducts_default = tool4;

// src/shared/prices/createPrice.ts
var import_zod5 = require("zod");
var createPricePrompt = (_context = {}) => `
This tool will create a price in Stripe. If a product has not already been specified, a product should be created first.

It takes three arguments:
- product (str): The ID of the product to create the price for.
- unit_amount (int): The unit amount of the price in cents.
- currency (str): The currency of the price.
`;
var createPrice = async (stripe, context, params) => {
  try {
    const price = await stripe.prices.create(
      params,
      context.account ? { stripeAccount: context.account } : void 0
    );
    return price;
  } catch (error) {
    return "Failed to create price";
  }
};
var createPriceParameters = (_context = {}) => import_zod5.z.object({
  product: import_zod5.z.string().describe("The ID of the product to create the price for."),
  unit_amount: import_zod5.z.number().int().describe("The unit amount of the price in cents."),
  currency: import_zod5.z.string().describe("The currency of the price.")
});
var createPriceAnnotations = () => ({
  destructiveHint: false,
  idempotentHint: false,
  openWorldHint: true,
  readOnlyHint: false,
  title: "Create price"
});
var tool5 = (context) => ({
  method: "create_price",
  name: "Create Price",
  description: createPricePrompt(context),
  parameters: createPriceParameters(context),
  annotations: createPriceAnnotations(),
  actions: {
    prices: {
      create: true
    }
  },
  execute: createPrice
});
var createPrice_default = tool5;

// src/shared/prices/listPrices.ts
var import_zod6 = require("zod");
var listPrices = async (stripe, context, params) => {
  try {
    const prices = await stripe.prices.list(
      params,
      context.account ? { stripeAccount: context.account } : void 0
    );
    return prices.data;
  } catch (error) {
    return "Failed to list prices";
  }
};
var listPricesParameters = (_context = {}) => import_zod6.z.object({
  product: import_zod6.z.string().optional().describe("The ID of the product to list prices for."),
  limit: import_zod6.z.number().int().min(1).max(100).optional().describe(
    "A limit on the number of objects to be returned. Limit can range between 1 and 100, and the default is 10."
  )
});
var listPricesAnnotations = () => ({
  destructiveHint: false,
  idempotentHint: true,
  openWorldHint: true,
  readOnlyHint: true,
  title: "List prices"
});
var listPricesPrompt = (_context = {}) => `
This tool will fetch a list of Prices from Stripe.

It takes two arguments.
- product (str, optional): The ID of the product to list prices for.
- limit (int, optional): The number of prices to return.
`;
var tool6 = (context) => ({
  method: "list_prices",
  name: "List Prices",
  description: listPricesPrompt(context),
  parameters: listPricesParameters(context),
  annotations: listPricesAnnotations(),
  actions: {
    prices: {
      read: true
    }
  },
  execute: listPrices
});
var listPrices_default = tool6;

// src/shared/paymentLinks/createPaymentLink.ts
var import_zod7 = require("zod");
var createPaymentLinkPrompt = (_context = {}) => `
This tool will create a payment link in Stripe.

It takes two arguments:
- price (str): The ID of the price to create the payment link for.
- quantity (int): The quantity of the product to include in the payment link.
- redirect_url (str, optional): The URL to redirect to after the payment is completed.
`;
var createPaymentLink = async (stripe, context, params) => {
  try {
    const paymentLink = await stripe.paymentLinks.create(
      {
        line_items: [{ price: params.price, quantity: params.quantity }],
        ...params.redirect_url ? {
          after_completion: {
            type: "redirect",
            redirect: {
              url: params.redirect_url
            }
          }
        } : void 0
      },
      context.account ? { stripeAccount: context.account } : void 0
    );
    return { id: paymentLink.id, url: paymentLink.url };
  } catch (error) {
    return "Failed to create payment link";
  }
};
var createPaymentLinkAnnotations = () => ({
  destructiveHint: false,
  idempotentHint: false,
  openWorldHint: true,
  readOnlyHint: false,
  title: "Create payment link"
});
var createPaymentLinkParameters = (_context = {}) => import_zod7.z.object({
  price: import_zod7.z.string().describe("The ID of the price to create the payment link for."),
  quantity: import_zod7.z.number().int().describe("The quantity of the product to include."),
  redirect_url: import_zod7.z.string().optional().describe("The URL to redirect to after the payment is completed.")
});
var tool7 = (context) => ({
  method: "create_payment_link",
  name: "Create Payment Link",
  description: createPaymentLinkPrompt(context),
  parameters: createPaymentLinkParameters(context),
  annotations: createPaymentLinkAnnotations(),
  actions: {
    paymentLinks: {
      create: true
    }
  },
  execute: createPaymentLink
});
var createPaymentLink_default = tool7;

// src/shared/invoices/createInvoice.ts
var import_zod8 = require("zod");
var createInvoicePrompt = (context = {}) => {
  const customerArg = context.customer ? `The customer is already set in the context: ${context.customer}.` : `- customer (str): The ID of the customer to create the invoice for.
`;
  return `
  This tool will create an invoice in Stripe.
  
  It takes ${context.customer ? "one" : "two"} argument${context.customer ? "" : "s"}:
  ${customerArg}
  - days_until_due (int, optional): The number of days until the invoice is due.
  `;
};
var createInvoiceParameters = (context = {}) => {
  const schema = import_zod8.z.object({
    customer: import_zod8.z.string().describe("The ID of the customer to create the invoice for."),
    days_until_due: import_zod8.z.number().int().optional().describe("The number of days until the invoice is due.")
  });
  if (context.customer) {
    return schema.omit({ customer: true });
  } else {
    return schema;
  }
};
var createInvoiceAnnotations = () => ({
  destructiveHint: false,
  idempotentHint: false,
  openWorldHint: true,
  readOnlyHint: false,
  title: "Create invoice"
});
var createInvoice = async (stripe, context, params) => {
  try {
    if (context.customer) {
      params.customer = context.customer;
    }
    const invoice = await stripe.invoices.create(
      {
        ...params,
        collection_method: "send_invoice"
      },
      context.account ? { stripeAccount: context.account } : void 0
    );
    return {
      id: invoice.id,
      url: invoice.hosted_invoice_url,
      customer: invoice.customer,
      status: invoice.status
    };
  } catch (error) {
    return "Failed to create invoice";
  }
};
var tool8 = (context) => ({
  method: "create_invoice",
  name: "Create Invoice",
  description: createInvoicePrompt(context),
  parameters: createInvoiceParameters(context),
  annotations: createInvoiceAnnotations(),
  actions: {
    invoices: {
      create: true
    }
  },
  execute: createInvoice
});
var createInvoice_default = tool8;

// src/shared/invoices/listInvoices.ts
var import_zod9 = require("zod");
var listInvoices = async (stripe, context, params) => {
  try {
    if (context.customer) {
      params.customer = context.customer;
    }
    const invoices = await stripe.invoices.list(
      params,
      context.account ? { stripeAccount: context.account } : void 0
    );
    return invoices.data;
  } catch (error) {
    return "Failed to list invoices";
  }
};
var listInvoicesAnnotations = () => ({
  destructiveHint: false,
  idempotentHint: true,
  openWorldHint: true,
  readOnlyHint: true,
  title: "List invoices"
});
var listInvoicesParameters = (context = {}) => {
  const schema = import_zod9.z.object({
    customer: import_zod9.z.string().optional().describe("The ID of the customer to list invoices for."),
    limit: import_zod9.z.number().int().min(1).max(100).optional().describe(
      "A limit on the number of objects to be returned. Limit can range between 1 and 100, and the default is 10."
    )
  });
  if (context.customer) {
    return schema.omit({ customer: true });
  } else {
    return schema;
  }
};
var listInvoicesPrompt = (context = {}) => {
  const customerArg = context.customer ? `The customer is already set in the context: ${context.customer}.` : `- customer (str, optional): The ID of the customer to list invoices for.
`;
  return `
This tool will fetch a list of Invoices from Stripe.

It takes ${context.customer ? "one" : "two"} argument${context.customer ? "" : "s"}:
${customerArg}
- limit (int, optional): The number of invoices to return.
`;
};
var tool9 = (context) => ({
  method: "list_invoices",
  name: "List Invoices",
  description: listInvoicesPrompt(context),
  parameters: listInvoicesParameters(context),
  annotations: listInvoicesAnnotations(),
  actions: {
    invoices: {
      read: true
    }
  },
  execute: listInvoices
});
var listInvoices_default = tool9;

// src/shared/invoiceItems/createInvoiceItem.ts
var import_zod10 = require("zod");
var createInvoiceItem = async (stripe, context, params) => {
  try {
    if (context.customer) {
      params.customer = context.customer;
    }
    const invoiceItem = await stripe.invoiceItems.create(
      // @ts-ignore
      params,
      context.account ? { stripeAccount: context.account } : void 0
    );
    return {
      id: invoiceItem.id,
      invoice: invoiceItem.invoice
    };
  } catch (error) {
    return "Failed to create invoice item";
  }
};
var createInvoiceItemAnnotations = () => ({
  destructiveHint: false,
  idempotentHint: false,
  openWorldHint: true,
  readOnlyHint: false,
  title: "Create invoice item"
});
var createInvoiceItemParameters = (context = {}) => {
  const schema = import_zod10.z.object({
    customer: import_zod10.z.string().describe("The ID of the customer to create the invoice item for."),
    price: import_zod10.z.string().describe("The ID of the price for the item."),
    invoice: import_zod10.z.string().describe("The ID of the invoice to create the item for.")
  });
  if (context.customer) {
    return schema.omit({ customer: true });
  } else {
    return schema;
  }
};
var createInvoiceItemPrompt = (context = {}) => {
  const customerArg = context.customer ? `The customer is already set in the context: ${context.customer}.` : `- customer (str): The ID of the customer to create the invoice item for.
`;
  return `
This tool will create an invoice item in Stripe.

It takes ${context.customer ? "two" : "three"} arguments'}:
${customerArg}
- price (str): The ID of the price to create the invoice item for.
- invoice (str): The ID of the invoice to create the invoice item for.
`;
};
var tool10 = (context) => ({
  method: "create_invoice_item",
  name: "Create Invoice Item",
  description: createInvoiceItemPrompt(context),
  parameters: createInvoiceItemParameters(context),
  annotations: createInvoiceItemAnnotations(),
  actions: {
    invoiceItems: {
      create: true
    }
  },
  execute: createInvoiceItem
});
var createInvoiceItem_default = tool10;

// src/shared/invoices/finalizeInvoice.ts
var import_zod11 = require("zod");
var finalizeInvoiceParameters = (_context = {}) => import_zod11.z.object({
  invoice: import_zod11.z.string().describe("The ID of the invoice to finalize.")
});
var finalizeInvoicePrompt = (_context = {}) => `
This tool will finalize an invoice in Stripe.

It takes one argument:
- invoice (str): The ID of the invoice to finalize.
`;
var finalizeInvoice = async (stripe, context, params) => {
  try {
    const invoice = await stripe.invoices.finalizeInvoice(
      params.invoice,
      context.account ? { stripeAccount: context.account } : void 0
    );
    return {
      id: invoice.id,
      url: invoice.hosted_invoice_url,
      customer: invoice.customer,
      status: invoice.status
    };
  } catch (error) {
    return "Failed to finalize invoice";
  }
};
var finalizeInvoiceAnnotations = () => ({
  destructiveHint: false,
  idempotentHint: true,
  openWorldHint: true,
  readOnlyHint: false,
  title: "Finalize invoice"
});
var tool11 = (context) => ({
  method: "finalize_invoice",
  name: "Finalize Invoice",
  description: finalizeInvoicePrompt(context),
  parameters: finalizeInvoiceParameters(context),
  annotations: finalizeInvoiceAnnotations(),
  actions: {
    invoices: {
      update: true
    }
  },
  execute: finalizeInvoice
});
var finalizeInvoice_default = tool11;

// src/shared/balance/retrieveBalance.ts
var import_zod12 = require("zod");
var retrieveBalancePrompt = (_context = {}) => `
This tool will retrieve the balance from Stripe. It takes no input.
`;
var retrieveBalanceParameters = (_context = {}) => import_zod12.z.object({});
var retrieveBalanceAnnotations = () => ({
  destructiveHint: false,
  idempotentHint: true,
  openWorldHint: true,
  readOnlyHint: true,
  title: "Retrieve balance"
});
var retrieveBalance = async (stripe, context, params) => {
  try {
    const balance = await stripe.balance.retrieve(
      params,
      context.account ? { stripeAccount: context.account } : void 0
    );
    return balance;
  } catch (error) {
    return "Failed to retrieve balance";
  }
};
var tool12 = (context) => ({
  method: "retrieve_balance",
  name: "Retrieve Balance",
  description: retrieveBalancePrompt(context),
  parameters: retrieveBalanceParameters(context),
  annotations: retrieveBalanceAnnotations(),
  actions: {
    balance: {
      read: true
    }
  },
  execute: retrieveBalance
});
var retrieveBalance_default = tool12;

// src/shared/coupons/listCoupons.ts
var import_zod13 = require("zod");
var listCouponsPrompt = (_context = {}) => `
This tool will fetch a list of Coupons from Stripe.

It takes one optional argument:
- limit (int, optional): The number of coupons to return.
`;
var listCouponsParameters = (_context = {}) => import_zod13.z.object({
  limit: import_zod13.z.number().int().min(1).max(100).optional().describe(
    "A limit on the number of objects to be returned. Limit can range between 1 and 100."
  )
});
var listCouponsAnnotations = () => ({
  destructiveHint: false,
  idempotentHint: true,
  openWorldHint: true,
  readOnlyHint: true,
  title: "List coupons"
});
var listCoupons = async (stripe, context, params) => {
  try {
    const coupons = await stripe.coupons.list(
      params,
      context.account ? { stripeAccount: context.account } : void 0
    );
    return coupons.data.map((coupon) => ({
      id: coupon.id,
      name: coupon.name,
      percent_off: coupon.percent_off,
      amount_off: coupon.amount_off,
      duration: coupon.duration
    }));
  } catch (error) {
    return "Failed to list coupons";
  }
};
var tool13 = (context) => ({
  method: "list_coupons",
  name: "List Coupons",
  description: listCouponsPrompt(context),
  parameters: listCouponsParameters(context),
  annotations: listCouponsAnnotations(),
  actions: {
    coupons: {
      read: true
    }
  },
  execute: listCoupons
});
var listCoupons_default = tool13;

// src/shared/coupons/createCoupon.ts
var import_zod14 = require("zod");
var createCouponPrompt = (_context = {}) => `
This tool will create a coupon in Stripe.


It takes several arguments:
- name (str): The name of the coupon.

Only use one of percent_off or amount_off, not both:
- percent_off (number, optional): The percentage discount to apply (between 0 and 100).
- amount_off (number, optional): The amount to subtract from an invoice (in cents).

Optional arguments for duration. Use if specific duration is desired, otherwise default to 'once'.
- duration (str, optional): How long the discount will last ('once', 'repeating', or 'forever'). Defaults to 'once'.
- duration_in_months (number, optional): The number of months the discount will last if duration is repeating.
`;
var createCouponParameters = (_context = {}) => import_zod14.z.object({
  name: import_zod14.z.string().describe(
    "Name of the coupon displayed to customers on invoices or receipts"
  ),
  percent_off: import_zod14.z.number().min(0).max(100).optional().describe(
    "A positive float larger than 0, and smaller or equal to 100, that represents the discount the coupon will apply (required if amount_off is not passed)"
  ),
  amount_off: import_zod14.z.number().describe(
    "A positive integer representing the amount to subtract from an invoice total (required if percent_off is not passed)"
  ),
  currency: import_zod14.z.string().optional().default("USD").describe(
    "Three-letter ISO code for the currency of the amount_off parameter (required if amount_off is passed). Infer based on the amount_off. For example, if a coupon is $2 off, set currency to be USD."
  ),
  duration: import_zod14.z.enum(["once", "repeating", "forever"]).default("once").optional().describe('How long the discount will last. Defaults to "once"'),
  duration_in_months: import_zod14.z.number().optional().describe(
    "The number of months the discount will last if duration is repeating"
  )
});
var createCouponAnnotations = () => ({
  destructiveHint: false,
  idempotentHint: false,
  openWorldHint: true,
  readOnlyHint: false,
  title: "Create coupon"
});
var createCoupon = async (stripe, context, params) => {
  try {
    const coupon = await stripe.coupons.create(
      params,
      context.account ? { stripeAccount: context.account } : void 0
    );
    return { id: coupon.id };
  } catch (error) {
    return `Failed to create coupon: ${error.message}`;
  }
};
var tool14 = (context) => ({
  method: "create_coupon",
  name: "Create Coupon",
  description: createCouponPrompt(context),
  parameters: createCouponParameters(context),
  annotations: createCouponAnnotations(),
  actions: {
    coupons: {
      create: true
    }
  },
  execute: createCoupon
});
var createCoupon_default = tool14;

// src/shared/refunds/createRefund.ts
var import_zod15 = require("zod");
var createRefundPrompt = (_context = {}) => `
This tool will refund a payment intent in Stripe.

It takes three arguments:
- payment_intent (str): The ID of the payment intent to refund.
- amount (int, optional): The amount to refund in cents.
- reason (str, optional): The reason for the refund.
`;
var createRefund = async (stripe, context, params) => {
  try {
    const refund = await stripe.refunds.create(
      params,
      context.account ? { stripeAccount: context.account } : void 0
    );
    return {
      id: refund.id,
      status: refund.status,
      amount: refund.amount
    };
  } catch (error) {
    return "Failed to create refund";
  }
};
var createRefundParameters = (_context = {}) => import_zod15.z.object({
  payment_intent: import_zod15.z.string().describe("The ID of the PaymentIntent to refund."),
  amount: import_zod15.z.number().int().optional().describe("The amount to refund in cents.")
});
var createRefundAnnotations = () => ({
  destructiveHint: false,
  idempotentHint: false,
  openWorldHint: true,
  readOnlyHint: false,
  title: "Create refund"
});
var tool15 = (context) => ({
  method: "create_refund",
  name: "Create Refund",
  description: createRefundPrompt(context),
  parameters: createRefundParameters(context),
  annotations: createRefundAnnotations(),
  actions: {
    refunds: {
      create: true
    }
  },
  execute: createRefund
});
var createRefund_default = tool15;

// src/shared/paymentIntents/listPaymentIntents.ts
var import_zod16 = require("zod");
var listPaymentIntentsPrompt = (context = {}) => {
  const customerArg = context.customer ? `The customer is already set in the context: ${context.customer}.` : `- customer (str, optional): The ID of the customer to list payment intents for.
`;
  return `
This tool will list payment intents in Stripe.

It takes ${context.customer ? "one" : "two"} argument${context.customer ? "" : "s"}:
${customerArg}
- limit (int, optional): The number of payment intents to return.
`;
};
var listPaymentIntents = async (stripe, context, params) => {
  try {
    if (context.customer) {
      params.customer = context.customer;
    }
    const paymentIntents = await stripe.paymentIntents.list(
      params,
      context.account ? { stripeAccount: context.account } : void 0
    );
    return paymentIntents.data;
  } catch (error) {
    return "Failed to list payment intents";
  }
};
var listPaymentIntentsAnnotations = () => ({
  destructiveHint: false,
  idempotentHint: true,
  openWorldHint: true,
  readOnlyHint: true,
  title: "List payment intents"
});
var listPaymentIntentsParameters = (context = {}) => {
  const schema = import_zod16.z.object({
    customer: import_zod16.z.string().optional().describe("The ID of the customer to list payment intents for."),
    limit: import_zod16.z.number().int().min(1).max(100).optional().describe(
      "A limit on the number of objects to be returned. Limit can range between 1 and 100."
    )
  });
  if (context.customer) {
    return schema.omit({ customer: true });
  } else {
    return schema;
  }
};
var tool16 = (context) => ({
  method: "list_payment_intents",
  name: "List Payment Intents",
  description: listPaymentIntentsPrompt(context),
  parameters: listPaymentIntentsParameters(context),
  annotations: listPaymentIntentsAnnotations(),
  actions: {
    paymentIntents: {
      read: true
    }
  },
  execute: listPaymentIntents
});
var listPaymentIntents_default = tool16;

// src/shared/subscriptions/listSubscriptions.ts
var import_zod17 = require("zod");
var listSubscriptions = async (stripe, context, params) => {
  try {
    if (context.customer) {
      params.customer = context.customer;
    }
    const subscriptions = await stripe.subscriptions.list(
      params,
      context.account ? { stripeAccount: context.account } : void 0
    );
    return subscriptions.data;
  } catch (error) {
    return "Failed to list subscriptions";
  }
};
var listSubscriptionsParameters = (context = {}) => {
  const schema = import_zod17.z.object({
    customer: import_zod17.z.string().optional().describe("The ID of the customer to list subscriptions for."),
    price: import_zod17.z.string().optional().describe("The ID of the price to list subscriptions for."),
    status: import_zod17.z.enum([
      "active",
      "past_due",
      "unpaid",
      "canceled",
      "incomplete",
      "incomplete_expired",
      "trialing",
      "all"
    ]).optional().describe("The status of the subscriptions to retrieve."),
    limit: import_zod17.z.number().int().min(1).max(100).optional().describe(
      "A limit on the number of objects to be returned. Limit can range between 1 and 100."
    )
  });
  if (context.customer) {
    return schema.omit({ customer: true });
  } else {
    return schema;
  }
};
var listSubscriptionsPrompt = (context = {}) => {
  const customerArg = context.customer ? `The customer is already set in the context: ${context.customer}.` : `- customer (str, optional): The ID of the customer to list subscriptions for.
`;
  return `
This tool will list all subscriptions in Stripe.

It takes ${context.customer ? "three" : "four"} arguments:
${customerArg}
- price (str, optional): The ID of the price to list subscriptions for.
- status (str, optional): The status of the subscriptions to list.
- limit (int, optional): The number of subscriptions to return.
`;
};
var listSubscriptionsAnnotations = () => ({
  destructiveHint: false,
  idempotentHint: true,
  openWorldHint: true,
  readOnlyHint: true,
  title: "List subscriptions"
});
var tool17 = (context) => ({
  method: "list_subscriptions",
  name: "List Subscriptions",
  description: listSubscriptionsPrompt(context),
  parameters: listSubscriptionsParameters(context),
  annotations: listSubscriptionsAnnotations(),
  actions: {
    subscriptions: {
      read: true
    }
  },
  execute: listSubscriptions
});
var listSubscriptions_default = tool17;

// src/shared/subscriptions/cancelSubscription.ts
var import_zod18 = require("zod");
var cancelSubscription = async (stripe, context, params) => {
  try {
    const { subscription: subscriptionId, ...cancelParams } = params;
    const subscription = await stripe.subscriptions.cancel(
      subscriptionId,
      cancelParams,
      context.account ? { stripeAccount: context.account } : void 0
    );
    return subscription;
  } catch (error) {
    return "Failed to cancel subscription";
  }
};
var cancelSubscriptionParameters = (_context = {}) => {
  return import_zod18.z.object({
    subscription: import_zod18.z.string().describe("The ID of the subscription to cancel.")
  });
};
var cancelSubscriptionPrompt = (_context = {}) => {
  return `
This tool will cancel a subscription in Stripe.

It takes the following arguments:
- subscription (str, required): The ID of the subscription to cancel.
`;
};
var cancelSubscriptionAnnotations = () => ({
  destructiveHint: true,
  idempotentHint: true,
  openWorldHint: true,
  readOnlyHint: false,
  title: "Cancel subscription"
});
var tool18 = (context) => ({
  method: "cancel_subscription",
  name: "Cancel Subscription",
  description: cancelSubscriptionPrompt(context),
  parameters: cancelSubscriptionParameters(context),
  annotations: cancelSubscriptionAnnotations(),
  actions: {
    subscriptions: {
      update: true
    }
  },
  execute: cancelSubscription
});
var cancelSubscription_default = tool18;

// src/shared/subscriptions/updateSubscription.ts
var import_zod19 = require("zod");
var updateSubscriptionPrompt = (_context = {}) => {
  return `This tool will update an existing subscription in Stripe. If changing an existing subscription item, the existing subscription item has to be set to deleted and the new one has to be added.
  
  It takes the following arguments:
  - subscription (str, required): The ID of the subscription to update.
  - proration_behavior (str, optional): Determines how to handle prorations when the subscription items change. Options: 'create_prorations', 'none', 'always_invoice', 'none_implicit'.
  - items (array, optional): A list of subscription items to update, add, or remove. Each item can have the following properties:
    - id (str, optional): The ID of the subscription item to modify.
    - price (str, optional): The ID of the price to switch to.
    - quantity (int, optional): The quantity of the plan to subscribe to.
    - deleted (bool, optional): Whether to delete this item.
  `;
};
var updateSubscription = async (stripe, context, params) => {
  try {
    const { subscription: subscriptionId, ...updateParams } = params;
    const subscription = await stripe.subscriptions.update(
      subscriptionId,
      updateParams,
      context.account ? { stripeAccount: context.account } : void 0
    );
    return subscription;
  } catch (error) {
    return "Failed to update subscription";
  }
};
var updateSubscriptionParameters = (_context = {}) => {
  return import_zod19.z.object({
    subscription: import_zod19.z.string().describe("The ID of the subscription to update."),
    proration_behavior: import_zod19.z.enum(["create_prorations", "none", "always_invoice", "none_implicit"]).optional().describe(
      "Determines how to handle prorations when the subscription items change."
    ),
    items: import_zod19.z.array(
      import_zod19.z.object({
        id: import_zod19.z.string().optional().describe("The ID of the subscription item to modify."),
        price: import_zod19.z.string().optional().describe("The ID of the price to switch to."),
        quantity: import_zod19.z.number().int().min(1).optional().describe("The quantity of the plan to subscribe to."),
        deleted: import_zod19.z.boolean().optional().describe("Whether to delete this item.")
      })
    ).optional().describe("A list of subscription items to update, add, or remove.")
  });
};
var updateSubscriptionAnnotations = () => ({
  destructiveHint: false,
  idempotentHint: false,
  openWorldHint: true,
  readOnlyHint: false,
  title: "Update subscription"
});
var tool19 = (context) => ({
  method: "update_subscription",
  name: "Update Subscription",
  description: updateSubscriptionPrompt(context),
  parameters: updateSubscriptionParameters(context),
  annotations: updateSubscriptionAnnotations(),
  actions: {
    subscriptions: {
      update: true
    }
  },
  execute: updateSubscription
});
var updateSubscription_default = tool19;

// src/shared/documentation/searchDocumentation.ts
var import_zod20 = require("zod");
var searchDocumentationPrompt = (_context = {}) => `
This tool will take in a user question about integrating with Stripe in their application, then search and retrieve relevant Stripe documentation to answer the question.

It takes two arguments:
- question (str): The user question to search an answer for in the Stripe documentation.
- language (str, optional): The programming language to search for in the the documentation.
`;
var searchDocumentationParameters = (_context = {}) => import_zod20.z.object({
  question: import_zod20.z.string().describe(
    "The user question about integrating with Stripe will be used to search the documentation."
  ),
  language: import_zod20.z.enum(["dotnet", "go", "java", "node", "php", "ruby", "python", "curl"]).optional().describe(
    "The programming language to search for in the the documentation."
  )
});
var searchDocumentationAnnotations = () => ({
  destructiveHint: false,
  idempotentHint: true,
  openWorldHint: true,
  readOnlyHint: true,
  title: "Search Stripe documentation"
});
var searchDocumentation = async (_stripe, context, params) => {
  try {
    const endpoint = "https://ai.stripe.com/search";
    const response = await fetch(endpoint, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "X-Requested-With": "fetch",
        "User-Agent": context.mode === "modelcontextprotocol" ? "stripe-mcp" : "stripe-agent-toolkit-typescript"
      },
      body: JSON.stringify(params)
    });
    if (!response.ok) {
      throw new Error(`HTTP error! Status: ${response.status}`);
    }
    const data = await response.json();
    return data?.sources;
  } catch (error) {
    return "Failed to search documentation";
  }
};
var tool20 = (context) => ({
  method: "search_stripe_documentation",
  name: "Search Stripe Documentation",
  description: searchDocumentationPrompt(context),
  parameters: searchDocumentationParameters(context),
  annotations: searchDocumentationAnnotations(),
  actions: {
    documentation: {
      read: true
    }
  },
  execute: searchDocumentation
});
var searchDocumentation_default = tool20;

// src/shared/disputes/listDisputes.ts
var import_zod21 = require("zod");
var listDisputesPrompt = (_context = {}) => `
This tool will fetch a list of disputes in Stripe.

It takes the following arguments:
- charge (string, optional): Only return disputes associated to the charge specified by this charge ID.
- payment_intent (string, optional): Only return disputes associated to the PaymentIntent specified by this PaymentIntent ID.
`;
var listDisputesParameters = (_context = {}) => import_zod21.z.object({
  charge: import_zod21.z.string().optional().describe(
    "Only return disputes associated to the charge specified by this charge ID."
  ),
  payment_intent: import_zod21.z.string().optional().describe(
    "Only return disputes associated to the PaymentIntent specified by this PaymentIntent ID."
  ),
  limit: import_zod21.z.number().int().min(1).max(100).default(10).optional().describe(
    "A limit on the number of objects to be returned. Limit can range between 1 and 100, and the default is 10."
  )
});
var listDisputesAnnotations = () => ({
  destructiveHint: false,
  idempotentHint: true,
  openWorldHint: true,
  readOnlyHint: true,
  title: "List disputes"
});
var listDisputes = async (stripe, context, params) => {
  try {
    const disputes = await stripe.disputes.list(
      params,
      context.account ? { stripeAccount: context.account } : void 0
    );
    return disputes.data.map((dispute) => ({ id: dispute.id }));
  } catch (error) {
    return "Failed to list disputes";
  }
};
var tool21 = (context) => ({
  method: "list_disputes",
  name: "List Disputes",
  description: listDisputesPrompt(context),
  parameters: listDisputesParameters(context),
  annotations: listDisputesAnnotations(),
  actions: {
    disputes: {
      read: true
    }
  },
  execute: listDisputes
});
var listDisputes_default = tool21;

// src/shared/disputes/updateDispute.ts
var import_zod22 = require("zod");
var updateDisputePrompt = (_context = {}) => `
When you receive a dispute, contacting your customer is always the best first step. If that doesn't work, you can submit evidence to help resolve the dispute in your favor. This tool helps.

It takes the following arguments:
- dispute (string): The ID of the dispute to update
- evidence (object, optional): Evidence to upload for the dispute.
    - cancellation_policy_disclosure (string)
    - cancellation_rebuttal (string)
    - duplicate_charge_explanation (string)
    - uncategorized_text (string, optional): Any additional evidence or statements.
- submit (boolean, optional): Whether to immediately submit evidence to the bank. If false, evidence is staged on the dispute.
`;
var updateDisputeParameters = (_context = {}) => import_zod22.z.object({
  dispute: import_zod22.z.string().describe("The ID of the dispute to update"),
  evidence: import_zod22.z.object({
    cancellation_policy_disclosure: import_zod22.z.string().max(2e4).optional().describe(
      "An explanation of how and when the customer was shown your refund policy prior to purchase."
    ),
    duplicate_charge_explanation: import_zod22.z.string().max(2e4).optional().describe(
      "An explanation of the difference between the disputed charge versus the prior charge that appears to be a duplicate."
    ),
    uncategorized_text: import_zod22.z.string().max(2e4).optional().describe("Any additional evidence or statements.")
  }).optional().describe(
    "Evidence to upload, to respond to a dispute. Updating any field in the hash will submit all fields in the hash for review."
  ),
  submit: import_zod22.z.boolean().optional().describe(
    "Whether to immediately submit evidence to the bank. If false, evidence is staged on the dispute."
  )
});
var updateDisputeAnnotations = () => ({
  destructiveHint: false,
  idempotentHint: false,
  openWorldHint: true,
  readOnlyHint: false,
  title: "Update dispute"
});
var updateDispute = async (stripe, context, params) => {
  try {
    const updateParams = {
      evidence: params.evidence,
      submit: params.submit
    };
    const updatedDispute = await stripe.disputes.update(
      params.dispute,
      updateParams,
      context.account ? { stripeAccount: context.account } : void 0
    );
    return { id: updatedDispute.id };
  } catch (error) {
    return "Failed to update dispute";
  }
};
var tool22 = (context) => ({
  method: "update_dispute",
  name: "Update Dispute",
  description: updateDisputePrompt(context),
  parameters: updateDisputeParameters(context),
  annotations: updateDisputeAnnotations(),
  actions: {
    disputes: {
      update: true
    }
  },
  execute: updateDispute
});
var updateDispute_default = tool22;

// src/shared/tools.ts
var tools = (context) => [
  createCustomer_default(context),
  listCustomers_default(context),
  createProduct_default(context),
  listProducts_default(context),
  createPrice_default(context),
  listPrices_default(context),
  createPaymentLink_default(context),
  createInvoice_default(context),
  listInvoices_default(context),
  createInvoiceItem_default(context),
  finalizeInvoice_default(context),
  retrieveBalance_default(context),
  createRefund_default(context),
  listPaymentIntents_default(context),
  listSubscriptions_default(context),
  cancelSubscription_default(context),
  updateSubscription_default(context),
  searchDocumentation_default(context),
  listCoupons_default(context),
  createCoupon_default(context),
  updateDispute_default(context),
  listDisputes_default(context)
];
var tools_default = tools;

// src/shared/api.ts
var TOOLKIT_HEADER = "stripe-agent-toolkit-typescript";
var MCP_HEADER = "stripe-mcp";
var StripeAPI = class {
  stripe;
  context;
  tools;
  constructor(secretKey, context) {
    const stripeClient = new import_stripe.default(secretKey, {
      appInfo: {
        name: context?.mode === "modelcontextprotocol" ? MCP_HEADER : TOOLKIT_HEADER,
        version: "0.7.11",
        url: "https://github.com/stripe/agent-toolkit"
      }
    });
    this.stripe = stripeClient;
    this.context = context || {};
    this.tools = tools_default(this.context);
  }
  async createMeterEvent({
    event,
    customer,
    value
  }) {
    await this.stripe.billing.meterEvents.create(
      {
        event_name: event,
        payload: {
          stripe_customer_id: customer,
          value
        }
      },
      this.context.account ? { stripeAccount: this.context.account } : void 0
    );
  }
  async run(method, arg) {
    const tool23 = this.tools.find((t) => t.method === method);
    if (tool23) {
      const output = JSON.stringify(
        await tool23.execute(this.stripe, this.context, arg)
      );
      return output;
    } else {
      throw new Error("Invalid method " + method);
    }
  }
};
var api_default = StripeAPI;

// src/shared/configuration.ts
var isToolAllowed = (tool23, configuration) => {
  return Object.keys(tool23.actions).every((resource) => {
    const permissions = tool23.actions[resource];
    return Object.keys(permissions).every((permission) => {
      return configuration.actions[resource]?.[permission] === true;
    });
  });
};

// src/langchain/toolkit.ts
var StripeAgentToolkit = class {
  _stripe;
  tools;
  constructor({
    secretKey,
    configuration
  }) {
    this._stripe = new api_default(secretKey, configuration.context);
    const context = configuration.context || {};
    const filteredTools = tools_default(context).filter(
      (tool23) => isToolAllowed(tool23, configuration)
    );
    this.tools = filteredTools.map(
      (tool23) => new tool_default(
        this._stripe,
        tool23.method,
        tool23.description,
        tool23.parameters
      )
    );
  }
  getTools() {
    return this.tools;
  }
};
var toolkit_default = StripeAgentToolkit;
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  StripeAgentToolkit
});
