// supermemory/resources/memories.js
"use strict";
// File generated from our OpenAPI spec by Stainless. See CONTRIBUTING.md for details.
Object.defineProperty(exports, "__esModule", { value: true });
exports.Memories = void 0;
const resource_1 = require("../core/resource.js");
const headers_1 = require("../internal/headers.js");
const uploads_1 = require("../internal/uploads.js");
const path_1 = require("../internal/utils/path.js");
class Memories extends resource_1.APIResource {
    /**
     * Update a document with any content type (text, url, file, etc.) and metadata
     *
     * @example
     * ```ts
     * const memory = await client.memories.update('id');
     * ```
     */
    update(id, body = {}, options) {
        return this._client.patch((0, path_1.path) `/v3/documents/${id}`, { body, ...options });
    }
    /**
     * Retrieves a paginated list of documents with their metadata and workflow status
     *
     * @example
     * ```ts
     * const memories = await client.memories.list();
     * ```
     */
    list(body = {}, options) {
        return this._client.post('/v3/documents/list', { body, ...options });
    }
    /**
     * Delete a document by ID or customId
     *
     * @example
     * ```ts
     * await client.memories.delete('id');
     * ```
     */
    delete(id, options) {
        return this._client.delete((0, path_1.path) `/v3/documents/${id}`, {
            ...options,
            headers: (0, headers_1.buildHeaders)([{ Accept: '*/*' }, options?.headers]),
        });
    }
    /**
     * Add a document with any content type (text, url, file, etc.) and metadata
     *
     * @example
     * ```ts
     * const response = await client.memories.add({
     *   content: 'content',
     * });
     * ```
     */
    add(body, options) {
        return this._client.post('/v3/documents', { body, ...options });
    }
    /**
     * Forget (soft delete) a memory entry. The memory is marked as forgotten but not
     * permanently deleted.
     *
     * @example
     * ```ts
     * const response = await client.memories.forget({
     *   containerTag: 'user_123',
     * });
     * ```
     */
    forget(body, options) {
        return this._client.delete('/v4/memories', { body, ...options });
    }
    /**
     * Get a document by ID
     *
     * @example
     * ```ts
     * const memory = await client.memories.get('id');
     * ```
     */
    get(id, options) {
        return this._client.get((0, path_1.path) `/v3/documents/${id}`, options);
    }
    /**
     * Update a memory by creating a new version. The original memory is preserved with
     * isLatest=false.
     *
     * @example
     * ```ts
     * const response = await client.memories.updateMemory({
     *   containerTag: 'user_123',
     *   newContent: 'John now prefers light mode',
     * });
     * ```
     */
    updateMemory(body, options) {
        return this._client.patch('/v4/memories', { body, ...options });
    }
    /**
     * Upload a file to be processed
     *
     * @example
     * ```ts
     * const response = await client.memories.uploadFile({
     *   file: fs.createReadStream('path/to/file'),
     * });
     * ```
     */
    uploadFile(body, options) {
        return this._client.post('/v3/documents/file', (0, uploads_1.multipartFormRequestOptions)({ body, ...options }, this._client));
    }
}
exports.Memories = Memories;
