// browser-use-sdk/dist/chunk-SKUINUZ4.js
// src/core/errors.ts
var BrowserUseError = class extends Error {
  statusCode;
  detail;
  constructor(statusCode, message, detail) {
    super(message);
    this.name = "BrowserUseError";
    this.statusCode = statusCode;
    this.detail = detail;
  }
};

// src/core/http.ts
var HttpClient = class {
  apiKey;
  baseUrl;
  maxRetries;
  timeout;
  fetchPromise;
  useApiKeyHeader;
  constructor(options) {
    this.apiKey = options.apiKey;
    this.baseUrl = options.baseUrl.replace(/\/+$/, "");
    this.maxRetries = options.maxRetries ?? 3;
    this.timeout = options.timeout ?? 3e4;
    this.fetchPromise = Promise.resolve(
      options.fetch ?? ((input, init) => fetch(input, init))
    );
    this.useApiKeyHeader = options.apiKey !== "";
  }
  async request(method, path, options) {
    const url = new URL(`${this.baseUrl}${path}`);
    if (options?.query) {
      for (const [key, value] of Object.entries(options.query)) {
        if (value !== void 0 && value !== null) {
          url.searchParams.set(key, String(value));
        }
      }
    }
    const headers = {};
    if (this.useApiKeyHeader) {
      headers["X-Browser-Use-API-Key"] = this.apiKey;
    }
    if (options?.body !== void 0) {
      headers["Content-Type"] = "application/json";
    }
    for (let attempt = 0; attempt <= this.maxRetries; attempt++) {
      if (attempt > 0) {
        const delay = Math.min(1e3 * 2 ** (attempt - 1), 1e4);
        await new Promise((resolve) => setTimeout(resolve, delay));
      }
      const controller = new AbortController();
      const timeoutId = options?.signal ? void 0 : setTimeout(() => controller.abort(), this.timeout);
      const signal = options?.signal ?? controller.signal;
      try {
        const fetchImpl = await this.fetchPromise;
        const response = await fetchImpl(url.toString(), {
          method,
          headers,
          body: options?.body !== void 0 ? JSON.stringify(options.body) : void 0,
          signal
        });
        clearTimeout(timeoutId);
        if (response.ok) {
          if (response.status === 204) {
            return void 0;
          }
          return await response.json();
        }
        const shouldRetry = response.status === 429 && attempt < this.maxRetries;
        if (shouldRetry) {
          continue;
        }
        let errorBody;
        try {
          errorBody = await response.json();
        } catch {
        }
        const raw = typeof errorBody === "object" && errorBody !== null ? "message" in errorBody ? errorBody.message : "detail" in errorBody ? errorBody.detail : void 0 : void 0;
        const message = raw === void 0 ? `HTTP ${response.status}` : typeof raw === "string" ? raw : JSON.stringify(raw);
        throw new BrowserUseError(response.status, message, errorBody);
      } catch (error) {
        clearTimeout(timeoutId);
        throw error;
      }
    }
    throw new Error("Unreachable: retry loop exhausted");
  }
  get(path, query) {
    return this.request("GET", path, { query });
  }
  post(path, body, query) {
    return this.request("POST", path, { body, query });
  }
  patch(path, body, query) {
    return this.request("PATCH", path, { body, query });
  }
  delete(path, query) {
    return this.request("DELETE", path, { query });
  }
};

// src/core/x402.ts
var X402_BASE_URL_DEFAULT = "https://x402.api.browser-use.com/api/v3";
var X402_BASE_URL_DEFAULT_V2 = "https://x402.api.browser-use.com/api/v2";
var MISSING_X402 = "x402 mode requires the optional peer deps. Install them with: npm install @x402/fetch @x402/evm viem  (or the equivalent for pnpm/yarn).";
async function x402ClientFromPrivateKey(privateKey) {
  let viem;
  let fetchPkg;
  let evmPkg;
  try {
    viem = await import("viem/accounts");
    fetchPkg = await import("@x402/fetch");
    evmPkg = await import("@x402/evm");
  } catch (e) {
    throw new Error(MISSING_X402);
  }
  const key = privateKey.startsWith("0x") ? privateKey : `0x${privateKey}`;
  const signer = viem.privateKeyToAccount(key);
  const client = new fetchPkg.x402Client();
  client.register("eip155:*", new evmPkg.ExactEvmScheme(signer));
  return client;
}
async function wrapFetchWithX402(fetch2, x402Client) {
  let fetchPkg;
  try {
    fetchPkg = await import("@x402/fetch");
  } catch (e) {
    throw new Error(MISSING_X402);
  }
  return fetchPkg.wrapFetchWithPayment(fetch2, x402Client);
}

export {
  BrowserUseError,
  HttpClient,
  X402_BASE_URL_DEFAULT,
  X402_BASE_URL_DEFAULT_V2,
  x402ClientFromPrivateKey,
  wrapFetchWithX402
};
