// browser-use-sdk/dist/chunk-4BAZFOC2.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _interopRequireWildcard(obj) { if (obj && obj.__esModule) { return obj; } else { var newObj = {}; if (obj != null) { for (var key in obj) { if (Object.prototype.hasOwnProperty.call(obj, key)) { newObj[key] = obj[key]; } } } newObj.default = obj; return newObj; } } function _nullishCoalesce(lhs, rhsFn) { if (lhs != null) { return lhs; } else { return rhsFn(); } } function _optionalChain(ops) { let lastAccessLHS = undefined; let value = ops[0]; let i = 1; while (i < ops.length) { const op = ops[i]; const fn = ops[i + 1]; i += 2; if ((op === 'optionalAccess' || op === 'optionalCall') && value == null) { return undefined; } if (op === 'access' || op === 'optionalAccess') { lastAccessLHS = value; value = fn(value); } else if (op === 'call' || op === 'optionalCall') { value = fn((...args) => value.call(lastAccessLHS, ...args)); lastAccessLHS = undefined; } } return value; }// src/core/errors.ts
var BrowserUseError = class extends Error {
  
  
  constructor(statusCode, message, detail) {
    super(message);
    this.name = "BrowserUseError";
    this.statusCode = statusCode;
    this.detail = detail;
  }
};

// src/core/http.ts
var HttpClient = class {
  
  
  
  
  
  
  constructor(options) {
    this.apiKey = options.apiKey;
    this.baseUrl = options.baseUrl.replace(/\/+$/, "");
    this.maxRetries = _nullishCoalesce(options.maxRetries, () => ( 3));
    this.timeout = _nullishCoalesce(options.timeout, () => ( 3e4));
    this.fetchPromise = Promise.resolve(
      _nullishCoalesce(options.fetch, () => ( ((input, init) => fetch(input, init))))
    );
    this.useApiKeyHeader = options.apiKey !== "";
  }
  async request(method, path, options) {
    const url = new URL(`${this.baseUrl}${path}`);
    if (_optionalChain([options, 'optionalAccess', _ => _.query])) {
      for (const [key, value] of Object.entries(options.query)) {
        if (value !== void 0 && value !== null) {
          url.searchParams.set(key, String(value));
        }
      }
    }
    const headers = {};
    if (this.useApiKeyHeader) {
      headers["X-Browser-Use-API-Key"] = this.apiKey;
    }
    if (_optionalChain([options, 'optionalAccess', _2 => _2.body]) !== void 0) {
      headers["Content-Type"] = "application/json";
    }
    for (let attempt = 0; attempt <= this.maxRetries; attempt++) {
      if (attempt > 0) {
        const delay = Math.min(1e3 * 2 ** (attempt - 1), 1e4);
        await new Promise((resolve) => setTimeout(resolve, delay));
      }
      const controller = new AbortController();
      const timeoutId = _optionalChain([options, 'optionalAccess', _3 => _3.signal]) ? void 0 : setTimeout(() => controller.abort(), this.timeout);
      const signal = _nullishCoalesce(_optionalChain([options, 'optionalAccess', _4 => _4.signal]), () => ( controller.signal));
      try {
        const fetchImpl = await this.fetchPromise;
        const response = await fetchImpl(url.toString(), {
          method,
          headers,
          body: _optionalChain([options, 'optionalAccess', _5 => _5.body]) !== void 0 ? JSON.stringify(options.body) : void 0,
          signal
        });
        clearTimeout(timeoutId);
        if (response.ok) {
          if (response.status === 204) {
            return void 0;
          }
          return await response.json();
        }
        const shouldRetry = response.status === 429 && attempt < this.maxRetries;
        if (shouldRetry) {
          continue;
        }
        let errorBody;
        try {
          errorBody = await response.json();
        } catch (e2) {
        }
        const raw = typeof errorBody === "object" && errorBody !== null ? "message" in errorBody ? errorBody.message : "detail" in errorBody ? errorBody.detail : void 0 : void 0;
        const message = raw === void 0 ? `HTTP ${response.status}` : typeof raw === "string" ? raw : JSON.stringify(raw);
        throw new BrowserUseError(response.status, message, errorBody);
      } catch (error) {
        clearTimeout(timeoutId);
        throw error;
      }
    }
    throw new Error("Unreachable: retry loop exhausted");
  }
  get(path, query) {
    return this.request("GET", path, { query });
  }
  post(path, body, query) {
    return this.request("POST", path, { body, query });
  }
  patch(path, body, query) {
    return this.request("PATCH", path, { body, query });
  }
  delete(path, query) {
    return this.request("DELETE", path, { query });
  }
};

// src/core/x402.ts
var X402_BASE_URL_DEFAULT = "https://x402.api.browser-use.com/api/v3";
var X402_BASE_URL_DEFAULT_V2 = "https://x402.api.browser-use.com/api/v2";
var MISSING_X402 = "x402 mode requires the optional peer deps. Install them with: npm install @x402/fetch @x402/evm viem  (or the equivalent for pnpm/yarn).";
async function x402ClientFromPrivateKey(privateKey) {
  let viem;
  let fetchPkg;
  let evmPkg;
  try {
    viem = await Promise.resolve().then(() => _interopRequireWildcard(require("viem/accounts")));
    fetchPkg = await Promise.resolve().then(() => _interopRequireWildcard(require("@x402/fetch")));
    evmPkg = await Promise.resolve().then(() => _interopRequireWildcard(require("@x402/evm")));
  } catch (e) {
    throw new Error(MISSING_X402);
  }
  const key = privateKey.startsWith("0x") ? privateKey : `0x${privateKey}`;
  const signer = viem.privateKeyToAccount(key);
  const client = new fetchPkg.x402Client();
  client.register("eip155:*", new evmPkg.ExactEvmScheme(signer));
  return client;
}
async function wrapFetchWithX402(fetch2, x402Client) {
  let fetchPkg;
  try {
    fetchPkg = await Promise.resolve().then(() => _interopRequireWildcard(require("@x402/fetch")));
  } catch (e) {
    throw new Error(MISSING_X402);
  }
  return fetchPkg.wrapFetchWithPayment(fetch2, x402Client);
}








exports.BrowserUseError = BrowserUseError; exports.HttpClient = HttpClient; exports.X402_BASE_URL_DEFAULT = X402_BASE_URL_DEFAULT; exports.X402_BASE_URL_DEFAULT_V2 = X402_BASE_URL_DEFAULT_V2; exports.x402ClientFromPrivateKey = x402ClientFromPrivateKey; exports.wrapFetchWithX402 = wrapFetchWithX402;
