// browser-use-sdk/dist/index.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _nullishCoalesce(lhs, rhsFn) { if (lhs != null) { return lhs; } else { return rhsFn(); } } async function _asyncNullishCoalesce(lhs, rhsFn) { if (lhs != null) { return lhs; } else { return await rhsFn(); } } function _optionalChain(ops) { let lastAccessLHS = undefined; let value = ops[0]; let i = 1; while (i < ops.length) { const op = ops[i]; const fn = ops[i + 1]; i += 2; if ((op === 'optionalAccess' || op === 'optionalCall') && value == null) { return undefined; } if (op === 'access' || op === 'optionalAccess') { lastAccessLHS = value; value = fn(value); } else if (op === 'call' || op === 'optionalCall') { value = fn((...args) => value.call(lastAccessLHS, ...args)); lastAccessLHS = undefined; } } return value; } var _class;





var _chunk4BAZFOC2cjs = require('./chunk-4BAZFOC2.cjs');

// src/v2/client.ts
var _zod = require('zod');

// src/v2/resources/billing.ts
var Billing = class {
  constructor(http) {
    this.http = http;
  }
  /** Get authenticated account billing information including credit balance. */
  account() {
    return this.http.get("/billing/account");
  }
};

// src/v2/resources/browsers.ts
var Browsers = class {
  constructor(http) {
    this.http = http;
  }
  /** Create a new browser session. */
  create(body = {}) {
    if (body.proxyCountryCode) {
      body = { ...body, proxyCountryCode: body.proxyCountryCode.toLowerCase() };
    }
    return this.http.post("/browsers", body);
  }
  /** List browser sessions with optional filtering. */
  list(params) {
    return this.http.get(
      "/browsers",
      params
    );
  }
  /** Get detailed browser session information. */
  get(sessionId) {
    return this.http.get(`/browsers/${sessionId}`);
  }
  /** Update a browser session (generic PATCH). */
  update(sessionId, body) {
    return this.http.patch(`/browsers/${sessionId}`, body);
  }
  /** Stop a browser session. */
  stop(sessionId) {
    return this.update(sessionId, { action: "stop" });
  }
  /** List files the browser downloaded to S3 during the session. */
  downloads(sessionId, params) {
    return this.http.get(
      `/browsers/${sessionId}/downloads`,
      params
    );
  }
};

// src/v2/resources/files.ts
var Files = class {
  constructor(http) {
    this.http = http;
  }
  /** Generate a presigned URL for uploading files to an agent session. */
  sessionUrl(sessionId, body) {
    return this.http.post(
      `/files/sessions/${sessionId}/presigned-url`,
      body
    );
  }
  /** Generate a presigned URL for uploading files to a browser session. */
  browserUrl(sessionId, body) {
    return this.http.post(
      `/files/browsers/${sessionId}/presigned-url`,
      body
    );
  }
  /** Get secure download URL for a task output file. */
  taskOutput(taskId, fileId) {
    return this.http.get(
      `/files/tasks/${taskId}/output-files/${fileId}`
    );
  }
};

// src/v2/resources/marketplace.ts
var Marketplace = class {
  constructor(http) {
    this.http = http;
  }
  /** List all public skills in the marketplace. */
  list(params) {
    return this.http.get(
      "/marketplace/skills",
      params
    );
  }
  /** Get details of a specific marketplace skill by slug. */
  get(skillSlug) {
    return this.http.get(`/marketplace/skills/${skillSlug}`);
  }
  /** Clone a public marketplace skill to your project. */
  clone(skillId) {
    return this.http.post(`/marketplace/skills/${skillId}/clone`);
  }
  /** Execute a marketplace skill. */
  execute(skillId, body) {
    return this.http.post(
      `/marketplace/skills/${skillId}/execute`,
      body
    );
  }
};

// src/v2/resources/profiles.ts
var Profiles = class {
  constructor(http) {
    this.http = http;
  }
  /** Create a new browser profile. */
  create(body) {
    return this.http.post("/profiles", body);
  }
  /** List profiles with pagination. */
  list(params) {
    return this.http.get("/profiles", params);
  }
  /** Get profile details. */
  get(profileId) {
    return this.http.get(`/profiles/${profileId}`);
  }
  /** Update a browser profile. */
  update(profileId, body) {
    return this.http.patch(`/profiles/${profileId}`, body);
  }
  /** Delete a browser profile. */
  delete(profileId) {
    return this.http.delete(`/profiles/${profileId}`);
  }
};

// src/v2/resources/sessions.ts
var Sessions = class {
  constructor(http) {
    this.http = http;
  }
  /** Create a new session. */
  create(body) {
    if (_optionalChain([body, 'optionalAccess', _ => _.proxyCountryCode])) {
      body = { ...body, proxyCountryCode: body.proxyCountryCode.toLowerCase() };
    }
    return this.http.post("/sessions", body);
  }
  /** List sessions with optional filtering. */
  list(params) {
    return this.http.get("/sessions", params);
  }
  /** Get detailed session information. */
  get(sessionId) {
    return this.http.get(`/sessions/${sessionId}`);
  }
  /** Update a session (generic PATCH). */
  update(sessionId, body) {
    return this.http.patch(`/sessions/${sessionId}`, body);
  }
  /** Stop a session and all its running tasks. */
  stop(sessionId) {
    return this.update(sessionId, { action: "stop" });
  }
  /** Delete a session with all its tasks. */
  delete(sessionId) {
    return this.http.delete(`/sessions/${sessionId}`);
  }
  /** Get public share information for a session. */
  getShare(sessionId) {
    return this.http.get(`/sessions/${sessionId}/public-share`);
  }
  /** Create or return existing public share for a session. */
  createShare(sessionId) {
    return this.http.post(`/sessions/${sessionId}/public-share`);
  }
  /** Remove public share for a session. */
  deleteShare(sessionId) {
    return this.http.delete(`/sessions/${sessionId}/public-share`);
  }
  /** Purge all session data (ZDR projects only). */
  purge(sessionId) {
    return this.http.post(`/sessions/${sessionId}/purge`);
  }
};

// src/v2/resources/skills.ts
var Skills = class {
  constructor(http) {
    this.http = http;
  }
  /** Create a new skill via automated generation. */
  create(body) {
    return this.http.post("/skills", body);
  }
  /** List all skills owned by the project. */
  list(params) {
    return this.http.get("/skills", params);
  }
  /** Get details of a specific skill. */
  get(skillId) {
    return this.http.get(`/skills/${skillId}`);
  }
  /** Delete a skill. */
  delete(skillId) {
    return this.http.delete(`/skills/${skillId}`);
  }
  /** Update skill metadata. */
  update(skillId, body) {
    return this.http.patch(`/skills/${skillId}`, body);
  }
  /** Cancel the current in-progress generation for a skill. */
  cancel(skillId) {
    return this.http.post(`/skills/${skillId}/cancel`);
  }
  /** Execute a skill with the provided parameters. */
  execute(skillId, body) {
    return this.http.post(`/skills/${skillId}/execute`, body);
  }
  /** Refine a skill based on feedback. */
  refine(skillId, body) {
    return this.http.post(`/skills/${skillId}/refine`, body);
  }
  /** Rollback to the previous version. */
  rollback(skillId) {
    return this.http.post(`/skills/${skillId}/rollback`);
  }
  /** List executions for a specific skill. */
  executions(skillId, params) {
    return this.http.get(
      `/skills/${skillId}/executions`,
      params
    );
  }
  /** Get presigned URL for downloading skill execution output. */
  executionOutput(skillId, executionId) {
    return this.http.get(
      `/skills/${skillId}/executions/${executionId}/output`
    );
  }
};

// src/v2/helpers.ts
var TERMINAL_STATUSES = /* @__PURE__ */ new Set(["finished", "stopped"]);
var TaskRun = (_class = class {
  
  
  
  
  
  __init() {this._taskId = null}
  __init2() {this._result = null}
  constructor(createPromise, tasks, schema, options) {;_class.prototype.__init.call(this);_class.prototype.__init2.call(this);
    this._createPromise = createPromise;
    this._tasks = tasks;
    this._schema = schema;
    this._timeout = _nullishCoalesce(_optionalChain([options, 'optionalAccess', _2 => _2.timeout]), () => ( 3e5));
    this._interval = _nullishCoalesce(_optionalChain([options, 'optionalAccess', _3 => _3.interval]), () => ( 2e3));
  }
  /** Task ID (available after creation resolves). */
  get taskId() {
    return this._taskId;
  }
  /** Full task result (available after awaiting or iterating to completion). */
  get result() {
    return this._result;
  }
  /** Enable `await client.run(...)` — polls status endpoint, returns TaskResult. */
  then(onFulfilled, onRejected) {
    return this._waitForOutput().then(onFulfilled, onRejected);
  }
  /** Enable `for await (const step of client.run(...))` — polls full task, yields new steps. */
  async *[Symbol.asyncIterator]() {
    const taskId = await this._ensureTaskId();
    let seen = 0;
    const deadline = Date.now() + this._timeout;
    while (Date.now() < deadline) {
      const task = await this._tasks.get(taskId);
      for (let i = seen; i < task.steps.length; i++) {
        yield task.steps[i];
      }
      seen = task.steps.length;
      if (TERMINAL_STATUSES.has(task.status)) {
        this._result = this._buildResult(task);
        return;
      }
      const remaining = deadline - Date.now();
      if (remaining <= 0) break;
      await new Promise((r) => setTimeout(r, Math.min(this._interval, remaining)));
    }
    throw new Error(`Task ${taskId} did not complete within ${this._timeout}ms`);
  }
  async _ensureTaskId() {
    if (this._taskId) return this._taskId;
    const created = await this._createPromise;
    this._taskId = created.id;
    return this._taskId;
  }
  /** Poll lightweight status endpoint until terminal, return TaskResult. */
  async _waitForOutput() {
    const taskId = await this._ensureTaskId();
    const deadline = Date.now() + this._timeout;
    while (Date.now() < deadline) {
      const status = await this._tasks.status(taskId);
      if (TERMINAL_STATUSES.has(status.status)) {
        const task = await this._tasks.get(taskId);
        this._result = this._buildResult(task);
        return this._result;
      }
      const remaining = deadline - Date.now();
      if (remaining <= 0) break;
      await new Promise((r) => setTimeout(r, Math.min(this._interval, remaining)));
    }
    throw new Error(`Task ${taskId} did not complete within ${this._timeout}ms`);
  }
  _buildResult(task) {
    const output = this._parseOutput(task.output);
    return { ...task, output };
  }
  _parseOutput(output) {
    if (output == null) return null;
    if (!this._schema) return output;
    return this._schema.parse(JSON.parse(output));
  }
}, _class);

// src/v2/resources/tasks.ts
var Tasks = class {
  constructor(http) {
    this.http = http;
  }
  /** Create and start a new AI agent task. */
  create(body) {
    return this.http.post("/tasks", body);
  }
  /** List tasks with optional filtering. */
  list(params) {
    return this.http.get("/tasks", params);
  }
  /** Get detailed task information. */
  get(taskId) {
    return this.http.get(`/tasks/${taskId}`);
  }
  /** Update a task (generic PATCH). */
  update(taskId, body) {
    return this.http.patch(`/tasks/${taskId}`, body);
  }
  /** Stop a running task. */
  stop(taskId) {
    return this.update(taskId, { action: "stop" });
  }
  /** Stop a running task and its associated browser session. */
  stopTaskAndSession(taskId) {
    return this.update(taskId, { action: "stop_task_and_session" });
  }
  /** Get lightweight task status (optimized for polling). */
  status(taskId) {
    return this.http.get(`/tasks/${taskId}/status`);
  }
  /** Get secure download URL for task execution logs. */
  logs(taskId) {
    return this.http.get(`/tasks/${taskId}/logs`);
  }
  /** Poll until a task reaches a terminal status, then return the full TaskView. */
  async wait(taskId, opts) {
    const timeout = _nullishCoalesce(_optionalChain([opts, 'optionalAccess', _4 => _4.timeout]), () => ( 3e5));
    const interval = _nullishCoalesce(_optionalChain([opts, 'optionalAccess', _5 => _5.interval]), () => ( 2e3));
    const deadline = Date.now() + timeout;
    while (Date.now() < deadline) {
      const status = await this.status(taskId);
      if (TERMINAL_STATUSES.has(status.status)) {
        return this.get(taskId);
      }
      const remaining = deadline - Date.now();
      if (remaining <= 0) break;
      await new Promise((r) => setTimeout(r, Math.min(interval, remaining)));
    }
    throw new Error(`Task ${taskId} did not complete within ${timeout}ms`);
  }
};

// src/v2/client.ts
var DEFAULT_BASE_URL = "https://api.browser-use.com/api/v2";
var BrowserUse = class {
  
  
  
  
  
  
  
  
  
  constructor(options = {}) {
    const x402PrivateKey = _nullishCoalesce(options.x402PrivateKey, () => ( process.env.BROWSER_USE_X402_PRIVATE_KEY));
    if (options.x402 || x402PrivateKey) {
      const topupKey = _nullishCoalesce(_nullishCoalesce(options.apiKey, () => ( process.env.BROWSER_USE_API_KEY)), () => ( ""));
      const fetchPromise = (async () => {
        const x402Client = await _asyncNullishCoalesce(options.x402, async () => ( await _chunk4BAZFOC2cjs.x402ClientFromPrivateKey.call(void 0, x402PrivateKey)));
        return _chunk4BAZFOC2cjs.wrapFetchWithX402.call(void 0, globalThis.fetch, x402Client);
      })();
      fetchPromise.catch(() => {
      });
      this.http = new (0, _chunk4BAZFOC2cjs.HttpClient)({
        apiKey: topupKey,
        baseUrl: _nullishCoalesce(options.baseUrl, () => ( _chunk4BAZFOC2cjs.X402_BASE_URL_DEFAULT_V2)),
        maxRetries: options.maxRetries,
        timeout: options.timeout,
        fetch: fetchPromise
      });
    } else {
      const apiKey = _nullishCoalesce(_nullishCoalesce(options.apiKey, () => ( process.env.BROWSER_USE_API_KEY)), () => ( ""));
      if (!apiKey) {
        throw new Error(
          "No credentials provided. Pass apiKey / set BROWSER_USE_API_KEY, or pass x402PrivateKey / set BROWSER_USE_X402_PRIVATE_KEY for pay-per-request access via USDC."
        );
      }
      this.http = new (0, _chunk4BAZFOC2cjs.HttpClient)({
        apiKey,
        baseUrl: _nullishCoalesce(options.baseUrl, () => ( DEFAULT_BASE_URL)),
        maxRetries: options.maxRetries,
        timeout: options.timeout
      });
    }
    this.billing = new Billing(this.http);
    this.tasks = new Tasks(this.http);
    this.sessions = new Sessions(this.http);
    this.files = new Files(this.http);
    this.profiles = new Profiles(this.http);
    this.browsers = new Browsers(this.http);
    this.skills = new Skills(this.http);
    this.marketplace = new Marketplace(this.http);
  }
  run(task, options) {
    const { schema, timeout, interval, ...rest } = _nullishCoalesce(options, () => ( {}));
    const body = { task, ...rest };
    if (schema) {
      body.structuredOutput = JSON.stringify(_zod.z.toJSONSchema(schema));
    }
    const promise = this.tasks.create(body);
    return new TaskRun(promise, this.tasks, schema, { timeout, interval });
  }
};












exports.Billing = Billing; exports.BrowserUse = BrowserUse; exports.BrowserUseError = _chunk4BAZFOC2cjs.BrowserUseError; exports.Browsers = Browsers; exports.Files = Files; exports.Marketplace = Marketplace; exports.Profiles = Profiles; exports.Sessions = Sessions; exports.Skills = Skills; exports.TaskRun = TaskRun; exports.Tasks = Tasks;
