// browser-use-sdk/dist/v3.cjs
"use strict";Object.defineProperty(exports, "__esModule", {value: true}); function _nullishCoalesce(lhs, rhsFn) { if (lhs != null) { return lhs; } else { return rhsFn(); } } async function _asyncNullishCoalesce(lhs, rhsFn) { if (lhs != null) { return lhs; } else { return await rhsFn(); } } function _optionalChain(ops) { let lastAccessLHS = undefined; let value = ops[0]; let i = 1; while (i < ops.length) { const op = ops[i]; const fn = ops[i + 1]; i += 2; if ((op === 'optionalAccess' || op === 'optionalCall') && value == null) { return undefined; } if (op === 'access' || op === 'optionalAccess') { lastAccessLHS = value; value = fn(value); } else if (op === 'call' || op === 'optionalCall') { value = fn((...args) => value.call(lastAccessLHS, ...args)); lastAccessLHS = undefined; } } return value; } var _class;





var _chunk4BAZFOC2cjs = require('./chunk-4BAZFOC2.cjs');

// src/v3/client.ts
var _zod = require('zod');

// src/v3/resources/billing.ts
var Billing = class {
  constructor(http) {
    this.http = http;
  }
  /** Get account billing information. */
  account() {
    return this.http.get("/billing/account");
  }
};

// src/v3/resources/browsers.ts
var Browsers = class {
  constructor(http) {
    this.http = http;
  }
  /** Create a standalone browser session. */
  create(body = {}) {
    return this.http.post("/browsers", body);
  }
  /** List browser sessions for the authenticated project. */
  list(params) {
    return this.http.get("/browsers", params);
  }
  /** Get browser session details. */
  get(sessionId) {
    return this.http.get(`/browsers/${sessionId}`);
  }
  /** Update a browser session (e.g. stop it). */
  update(sessionId, body) {
    return this.http.patch(`/browsers/${sessionId}`, body);
  }
  /** Stop a browser session. Convenience wrapper around update. */
  stop(sessionId) {
    return this.update(sessionId, { action: "stop" });
  }
  /** List files the browser downloaded to S3 during the session. */
  downloads(sessionId, params) {
    return this.http.get(
      `/browsers/${sessionId}/downloads`,
      params
    );
  }
};

// src/v3/resources/profiles.ts
var Profiles = class {
  constructor(http) {
    this.http = http;
  }
  /** Create a browser profile. */
  create(body) {
    return this.http.post("/profiles", body);
  }
  /** List profiles for the authenticated project. */
  list(params) {
    return this.http.get("/profiles", params);
  }
  /** Get profile details. */
  get(profileId) {
    return this.http.get(`/profiles/${profileId}`);
  }
  /** Update a profile. */
  update(profileId, body) {
    return this.http.patch(`/profiles/${profileId}`, body);
  }
  /** Delete a profile. */
  delete(profileId) {
    return this.http.delete(`/profiles/${profileId}`);
  }
};

// src/v3/resources/sessions.ts
var Sessions = class {
  constructor(http, options = {}) {
    this.http = http;
    this.options = options;
  }
  /** Create a session and optionally dispatch a task. */
  create(body) {
    const requestBody = { ..._nullishCoalesce(body, () => ( {})) };
    if (this.options.useOwnKey !== void 0 && requestBody.useOwnKey === void 0) {
      requestBody.useOwnKey = this.options.useOwnKey;
    }
    return this.http.post("/sessions", requestBody);
  }
  /** List sessions for the authenticated project. */
  list(params) {
    return this.http.get("/sessions", params);
  }
  /** Get session details. */
  get(sessionId) {
    return this.http.get(`/sessions/${sessionId}`);
  }
  /** Stop a session or the running task. */
  stop(sessionId, body) {
    return this.http.post(`/sessions/${sessionId}/stop`, body);
  }
  /** Soft-delete a session. */
  delete(sessionId) {
    return this.http.delete(`/sessions/${sessionId}`);
  }
  /** List messages for a session with cursor-based pagination. */
  messages(sessionId, params) {
    return this.http.get(
      `/sessions/${sessionId}/messages`,
      params
    );
  }
  /**
   * Poll until recording URLs are available. Returns presigned MP4 URLs.
   *
   * Returns an empty array if no recording was produced (e.g. the agent
   * answered without opening a browser, or recording was not enabled).
   */
  async waitForRecording(sessionId, options) {
    const timeout = _nullishCoalesce(_optionalChain([options, 'optionalAccess', _ => _.timeout]), () => ( 15e3));
    const interval = _nullishCoalesce(_optionalChain([options, 'optionalAccess', _2 => _2.interval]), () => ( 2e3));
    const deadline = Date.now() + timeout;
    while (Date.now() < deadline) {
      const session = await this.get(sessionId);
      if (_optionalChain([session, 'access', _3 => _3.recordingUrls, 'optionalAccess', _4 => _4.length])) return session.recordingUrls;
      const remaining = deadline - Date.now();
      if (remaining <= 0) break;
      await new Promise((r) => setTimeout(r, Math.min(interval, remaining)));
    }
    return [];
  }
};

// src/v3/resources/workspaces.ts
var _fs = require('fs');
var _path = require('path');
function safeJoin(base, untrusted) {
  const baseResolved = _path.resolve.call(void 0, base) + "/";
  const resolved = _path.resolve.call(void 0, base, untrusted);
  if (resolved !== _path.resolve.call(void 0, base) && !resolved.startsWith(baseResolved)) {
    throw new Error(`Path traversal detected: ${untrusted}`);
  }
  return resolved;
}
var MIME_TYPES = {
  ".csv": "text/csv",
  ".json": "application/json",
  ".txt": "text/plain",
  ".md": "text/markdown",
  ".html": "text/html",
  ".xml": "application/xml",
  ".yaml": "application/yaml",
  ".yml": "application/yaml",
  ".pdf": "application/pdf",
  ".png": "image/png",
  ".jpg": "image/jpeg",
  ".jpeg": "image/jpeg",
  ".gif": "image/gif",
  ".webp": "image/webp",
  ".svg": "image/svg+xml",
  ".mp4": "video/mp4",
  ".mp3": "audio/mpeg",
  ".wav": "audio/wav",
  ".zip": "application/zip",
  ".gz": "application/gzip",
  ".tar": "application/x-tar",
  ".xlsx": "application/vnd.openxmlformats-officedocument.spreadsheetml.sheet",
  ".xls": "application/vnd.ms-excel",
  ".docx": "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
  ".doc": "application/msword",
  ".pptx": "application/vnd.openxmlformats-officedocument.presentationml.presentation"
};
function guessContentType(path) {
  return _nullishCoalesce(MIME_TYPES[_path.extname.call(void 0, path).toLowerCase()], () => ( "application/octet-stream"));
}
var Workspaces = class {
  constructor(http) {
    this.http = http;
  }
  /** List workspaces for the authenticated project. */
  list(params) {
    return this.http.get("/workspaces", params);
  }
  /** Create a new workspace. */
  create(body) {
    return this.http.post("/workspaces", body);
  }
  /** Get workspace details. */
  get(workspaceId) {
    return this.http.get(`/workspaces/${workspaceId}`);
  }
  /** Update a workspace. */
  update(workspaceId, body) {
    return this.http.patch(`/workspaces/${workspaceId}`, body);
  }
  /** Delete a workspace and its data. */
  delete(workspaceId) {
    return this.http.delete(`/workspaces/${workspaceId}`);
  }
  /** List files in a workspace. */
  files(workspaceId, params) {
    return this.http.get(
      `/workspaces/${workspaceId}/files`,
      params
    );
  }
  /** Get presigned upload URLs for workspace files. */
  uploadFiles(workspaceId, body, query) {
    return this.http.post(
      `/workspaces/${workspaceId}/files/upload`,
      body,
      query
    );
  }
  /** Delete a file from a workspace. */
  deleteFile(workspaceId, path) {
    return this.http.delete(`/workspaces/${workspaceId}/files`, { path });
  }
  /** Get storage usage for a workspace. */
  size(workspaceId) {
    return this.http.get(`/workspaces/${workspaceId}/size`);
  }
  /**
   * Upload local files to a workspace. Returns the list of remote paths.
   *
   * ```ts
   * await client.workspaces.upload(wsId, "data.csv", "config.json");
   * await client.workspaces.upload(wsId, "data.csv", { prefix: "uploads/" });
   * ```
   */
  async upload(workspaceId, ...args) {
    let prefix;
    const paths = [];
    for (const arg of args) {
      if (typeof arg === "string") {
        paths.push(arg);
      } else if (typeof arg === "object" && arg !== null) {
        prefix = arg.prefix;
      }
    }
    if (paths.length === 0) {
      throw new Error("At least one file path is required");
    }
    const items = paths.map((p) => ({
      name: _path.basename.call(void 0, p),
      contentType: guessContentType(p),
      size: _fs.statSync.call(void 0, p).size
    }));
    const resp = await this.uploadFiles(workspaceId, { files: items }, prefix ? { prefix } : void 0);
    for (let i = 0; i < paths.length; i++) {
      const body = _fs.readFileSync.call(void 0, paths[i]);
      const res = await fetch(resp.files[i].uploadUrl, {
        method: "PUT",
        headers: { "Content-Type": items[i].contentType },
        body
      });
      if (!res.ok) throw new Error(`Upload failed: ${res.status} ${res.statusText}`);
    }
    return resp.files.map((f) => f.path);
  }
  /**
   * Download a single file from a workspace. Returns the local path.
   *
   * ```ts
   * const local = await client.workspaces.download(wsId, "uploads/data.csv", { to: "./data.csv" });
   * ```
   */
  async download(workspaceId, path, options) {
    const fileList = await this.files(workspaceId, { prefix: path, includeUrls: true });
    const match = _optionalChain([fileList, 'access', _5 => _5.files, 'optionalAccess', _6 => _6.find, 'call', _7 => _7((f) => f.path === path)]);
    if (!match) throw new Error(`File not found in workspace: ${path}`);
    const dest = _nullishCoalesce(_optionalChain([options, 'optionalAccess', _8 => _8.to]), () => ( _path.basename.call(void 0, match.path)));
    _fs.mkdirSync.call(void 0, _path.dirname.call(void 0, dest), { recursive: true });
    const resp = await fetch(match.url);
    if (!resp.ok) throw new Error(`Download failed: ${resp.status}`);
    _fs.writeFileSync.call(void 0, dest, Buffer.from(await resp.arrayBuffer()));
    return dest;
  }
  /**
   * Download all files from a workspace. Returns list of local paths.
   *
   * ```ts
   * const paths = await client.workspaces.downloadAll(wsId, { to: "./output" });
   * ```
   */
  async downloadAll(workspaceId, options) {
    const destDir = _nullishCoalesce(_optionalChain([options, 'optionalAccess', _9 => _9.to]), () => ( "."));
    _fs.mkdirSync.call(void 0, destDir, { recursive: true });
    const results = [];
    let cursor;
    do {
      const fileList = await this.files(workspaceId, {
        prefix: _optionalChain([options, 'optionalAccess', _10 => _10.prefix]),
        includeUrls: true,
        cursor
      });
      for (const f of _nullishCoalesce(fileList.files, () => ( []))) {
        const local = safeJoin(destDir, f.path);
        _fs.mkdirSync.call(void 0, _path.dirname.call(void 0, local), { recursive: true });
        const resp = await fetch(f.url);
        if (!resp.ok) throw new Error(`Download failed for ${f.path}: ${resp.status}`);
        _fs.writeFileSync.call(void 0, local, Buffer.from(await resp.arrayBuffer()));
        results.push(local);
      }
      cursor = fileList.hasMore ? _nullishCoalesce(fileList.nextCursor, () => ( void 0)) : void 0;
    } while (cursor);
    return results;
  }
};

// src/v3/helpers.ts
var TERMINAL_STATUSES = /* @__PURE__ */ new Set(["idle", "stopped", "timed_out", "error"]);
var SessionRun = (_class = class {
  
  
  
  
  
  
  __init() {this._sessionId = null}
  __init2() {this._result = null}
  constructor(createPromise, sessions, schema, options) {;_class.prototype.__init.call(this);_class.prototype.__init2.call(this);
    this._createPromise = createPromise;
    this._sessions = sessions;
    this._schema = schema;
    this._timeout = _nullishCoalesce(_optionalChain([options, 'optionalAccess', _11 => _11.timeout]), () => ( 144e5));
    this._interval = _nullishCoalesce(_optionalChain([options, 'optionalAccess', _12 => _12.interval]), () => ( 2e3));
    this._options = options;
  }
  /** The session ID, available after task creation resolves. */
  get sessionId() {
    return this._sessionId;
  }
  /** The full SessionResult, available after polling completes. */
  get result() {
    return this._result;
  }
  /** Enable `await client.run(...)` — polls until terminal, returns SessionResult. */
  then(onFulfilled, onRejected) {
    return this._waitForOutput().then(onFulfilled, onRejected);
  }
  async _ensureSessionId() {
    if (this._sessionId) return this._sessionId;
    const created = await this._createPromise;
    this._sessionId = created.id;
    return this._sessionId;
  }
  /** Poll session until terminal, return SessionResult. */
  async _waitForOutput() {
    const sessionId = await this._ensureSessionId();
    const deadline = Date.now() + this._timeout;
    while (Date.now() < deadline) {
      const session = await this._sessions.get(sessionId);
      if (TERMINAL_STATUSES.has(session.status)) {
        const { output, ...rest } = session;
        const parsed = this._parseOutput(output);
        this._result = { ...rest, output: parsed };
        return this._result;
      }
      const remaining = deadline - Date.now();
      if (remaining <= 0) break;
      await new Promise(
        (r) => setTimeout(r, Math.min(this._interval, remaining))
      );
    }
    throw new Error(
      `Session ${sessionId} did not complete within ${this._timeout}ms`
    );
  }
  /**
   * Enable `for await (const msg of client.run(...))` — yields messages as they appear.
   * After iteration, `.result` contains the final SessionResult.
   */
  async *[Symbol.asyncIterator]() {
    const sessionId = await this._ensureSessionId();
    let cursor = _optionalChain([this, 'access', _13 => _13._options, 'optionalAccess', _14 => _14._startCursor]);
    const deadline = Date.now() + this._timeout;
    while (Date.now() < deadline) {
      const resp = await this._sessions.messages(sessionId, { after: cursor, limit: 100 });
      for (const msg of resp.messages) {
        yield msg;
        cursor = msg.id;
      }
      const session = await this._sessions.get(sessionId);
      if (TERMINAL_STATUSES.has(session.status)) {
        while (true) {
          const page = await this._sessions.messages(sessionId, { after: cursor, limit: 100 });
          if (!page.messages.length) break;
          for (const msg of page.messages) {
            yield msg;
            cursor = msg.id;
          }
        }
        const { output, ...rest } = session;
        this._result = { ...rest, output: this._parseOutput(output) };
        return;
      }
      const remaining = deadline - Date.now();
      if (remaining <= 0) break;
      await new Promise((r) => setTimeout(r, Math.min(this._interval, remaining)));
    }
    throw new Error(`Session ${sessionId} did not complete within ${this._timeout}ms`);
  }
  _parseOutput(output) {
    if (output == null) return null;
    if (!this._schema) return output;
    const raw = typeof output === "string" ? JSON.parse(output) : output;
    return this._schema.parse(raw);
  }
}, _class);

// src/v3/client.ts
var DEFAULT_BASE_URL = "https://api.browser-use.com/api/v3";
var BrowserUse = class {
  
  
  
  
  
  
  constructor(options = {}) {
    const x402PrivateKey = _nullishCoalesce(options.x402PrivateKey, () => ( process.env.BROWSER_USE_X402_PRIVATE_KEY));
    if (options.x402 || x402PrivateKey) {
      const topupKey = _nullishCoalesce(_nullishCoalesce(options.apiKey, () => ( process.env.BROWSER_USE_API_KEY)), () => ( ""));
      const fetchPromise = (async () => {
        const x402Client = await _asyncNullishCoalesce(options.x402, async () => ( await _chunk4BAZFOC2cjs.x402ClientFromPrivateKey.call(void 0, x402PrivateKey)));
        return _chunk4BAZFOC2cjs.wrapFetchWithX402.call(void 0, globalThis.fetch, x402Client);
      })();
      fetchPromise.catch(() => {
      });
      this.http = new (0, _chunk4BAZFOC2cjs.HttpClient)({
        apiKey: topupKey,
        baseUrl: _nullishCoalesce(options.baseUrl, () => ( _chunk4BAZFOC2cjs.X402_BASE_URL_DEFAULT)),
        maxRetries: options.maxRetries,
        timeout: options.timeout,
        fetch: fetchPromise
      });
    } else {
      const apiKey = _nullishCoalesce(_nullishCoalesce(options.apiKey, () => ( process.env.BROWSER_USE_API_KEY)), () => ( ""));
      if (!apiKey) {
        throw new Error(
          "No credentials provided. Pass apiKey / set BROWSER_USE_API_KEY, or pass x402PrivateKey / set BROWSER_USE_X402_PRIVATE_KEY for pay-per-request access via USDC."
        );
      }
      this.http = new (0, _chunk4BAZFOC2cjs.HttpClient)({
        apiKey,
        baseUrl: _nullishCoalesce(options.baseUrl, () => ( DEFAULT_BASE_URL)),
        maxRetries: options.maxRetries,
        timeout: options.timeout
      });
    }
    this.billing = new Billing(this.http);
    this.browsers = new Browsers(this.http);
    this.profiles = new Profiles(this.http);
    this.sessions = new Sessions(this.http, { useOwnKey: options.useOwnKey });
    this.workspaces = new Workspaces(this.http);
  }
  run(task, options) {
    const { schema, timeout, interval, ...rest } = _nullishCoalesce(options, () => ( {}));
    const body = { task, ...rest };
    if (body.proxyCountryCode) {
      body.proxyCountryCode = body.proxyCountryCode.toLowerCase();
    }
    if (schema) {
      if (typeof schema !== "object" || !("_zod" in schema || "_def" in schema)) {
        throw new Error(
          "schema must be a Zod schema (e.g. z.object({...})). Make sure you are using Zod v4: npm install zod@4"
        );
      }
      body.outputSchema = _zod.z.toJSONSchema(schema);
    }
    if (body.sessionId && body.keepAlive === void 0) {
      body.keepAlive = true;
    }
    if (body.sessionId) {
      const sid = body.sessionId;
      const sessions = this.sessions;
      let startCursor;
      const promise2 = sessions.messages(sid, { limit: 1 }).then((resp) => {
        const last = resp.messages[resp.messages.length - 1];
        startCursor = _optionalChain([last, 'optionalAccess', _15 => _15.id]);
      }).then(() => sessions.create(body));
      return new SessionRun(promise2, this.sessions, schema, {
        timeout,
        interval,
        get _startCursor() {
          return startCursor;
        }
      });
    }
    const promise = this.sessions.create(body);
    return new SessionRun(promise, this.sessions, schema, { timeout, interval });
  }
};









exports.Billing = Billing; exports.BrowserUse = BrowserUse; exports.BrowserUseError = _chunk4BAZFOC2cjs.BrowserUseError; exports.Browsers = Browsers; exports.Profiles = Profiles; exports.SessionRun = SessionRun; exports.Sessions = Sessions; exports.Workspaces = Workspaces;
