// agentmail/dist/cjs/Client.js
"use strict";
/**
 * This file was auto-generated by Fern from our API Definition.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.AgentMailClient = void 0;
const core = __importStar(require("./core/index.js"));
const headers_js_1 = require("./core/headers.js");
const Client_js_1 = require("./api/resources/inboxes/client/Client.js");
const Client_js_2 = require("./api/resources/webhooks/client/Client.js");
const Client_js_3 = require("./api/resources/domains/client/Client.js");
const Client_js_4 = require("./api/resources/drafts/client/Client.js");
const Client_js_5 = require("./api/resources/metrics/client/Client.js");
const Client_js_6 = require("./api/resources/threads/client/Client.js");
const Client_js_7 = require("./api/resources/websockets/client/Client.js");
class AgentMailClient {
    constructor(_options = {}) {
        this._options = Object.assign(Object.assign({}, _options), { headers: (0, headers_js_1.mergeHeaders)({
                "X-Fern-Language": "JavaScript",
                "X-Fern-SDK-Name": "agentmail",
                "X-Fern-SDK-Version": "0.0.51",
                "User-Agent": "agentmail/0.0.51",
                "X-Fern-Runtime": core.RUNTIME.type,
                "X-Fern-Runtime-Version": core.RUNTIME.version,
            }, _options === null || _options === void 0 ? void 0 : _options.headers) });
    }
    get inboxes() {
        var _a;
        return ((_a = this._inboxes) !== null && _a !== void 0 ? _a : (this._inboxes = new Client_js_1.Inboxes(this._options)));
    }
    get webhooks() {
        var _a;
        return ((_a = this._webhooks) !== null && _a !== void 0 ? _a : (this._webhooks = new Client_js_2.Webhooks(this._options)));
    }
    get domains() {
        var _a;
        return ((_a = this._domains) !== null && _a !== void 0 ? _a : (this._domains = new Client_js_3.Domains(this._options)));
    }
    get drafts() {
        var _a;
        return ((_a = this._drafts) !== null && _a !== void 0 ? _a : (this._drafts = new Client_js_4.Drafts(this._options)));
    }
    get metrics() {
        var _a;
        return ((_a = this._metrics) !== null && _a !== void 0 ? _a : (this._metrics = new Client_js_5.Metrics(this._options)));
    }
    get threads() {
        var _a;
        return ((_a = this._threads) !== null && _a !== void 0 ? _a : (this._threads = new Client_js_6.Threads(this._options)));
    }
    get websockets() {
        var _a;
        return ((_a = this._websockets) !== null && _a !== void 0 ? _a : (this._websockets = new Client_js_7.Websockets(this._options)));
    }
}
exports.AgentMailClient = AgentMailClient;
